src/transformStateWithClones.js

```diff
@@ -8,6 +8,48 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  const stateCopy = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        stateHistory.push(addProperties(stateCopy, action.extraData));
+        break;
+      case 'removeProperties':
+        stateHistory.push(removeProperties(stateCopy, action.keysToRemove));
+        break;
+      case 'clear':
+        stateHistory.push(clearState(stateCopy));
+        break;
+    }
```

Comment:
Consider using a default case in your switch statement for error handling. This will ensure that if an action type is not recognized, you can handle it appropriately, perhaps by throwing an error or ignoring the action.